File: src/main/java/dev/psygamer/wireframe/util/collection/FreezableSet.java
package dev.psygamer.wireframe.util.collection;

import dev.psygamer.wireframe.util.helper.ICloneable;
import dev.psygamer.wireframe.util.helper.IFreezable;

import com.google.common.collect.ImmutableSet;

import java.util.Set;
import java.util.Iterator;
import java.util.Collection;

public abstract class FreezableSet <E> implements Set<E>, IFreezable, ICloneable<FreezableSet<E>> {
	
	private final Set<E> set;
	private volatile boolean frozen = false;
	
	protected FreezableSet(final Set<E> set) {
		this.set = set;
	}
	
	@Override
	public int size() {
		return this.set.size();
	}
	
	@Override
	public boolean isEmpty() {
		return this.set.isEmpty();
	}
	
	@Override
	public boolean contains(final Object o) {
		return this.set.contains(o);
	}
	
	@Override
	public Iterator<E> iterator() {
		return this.set.iterator();
	}
	
	@Override
	public Object[] toArray() {
		return this.set.toArray();
	}
	
	@Override
	@SuppressWarnings("SuspiciousToArrayCall")
	public <T> T[] toArray(final T[] a) {
		return this.set.toArray(a);
	}
	
	@Override
	public boolean add(final E e) {
		IFreezable.throwIfFrozen(this);
		return this.set.add(e);
	}
	
	@Override
	public boolean remove(final Object o) {
		IFreezable.throwIfFrozen(this);
		return this.set.remove(o);
	}
	
	@Override
	public boolean containsAll(final Collection<?> c) {
		return this.set.containsAll(c);
	}
	
	@Override
	public boolean addAll(final Collection<? extends E> c) {
		IFreezable.throwIfFrozen(this);
		return this.set.addAll(c);
	}
	
	@Override
	public boolean retainAll(final Collection<?> c) {
		IFreezable.throwIfFrozen(this);
		return this.set.retainAll(c);
	}
	
	@Override
	public boolean removeAll(final Collection<?> c) {
		IFreezable.throwIfFrozen(this);
		return this.set.removeAll(c);
	}
	
	@Override
	public void clear() {
		IFreezable.throwIfFrozen(this);
		this.set.clear();
	}
	
	@Override
	public void freeze() {
		this.frozen = true;
	}
	
	@Override
	public boolean isFrozen() {
		return this.frozen;
	}
	
	public ImmutableSet<E> toImmutable() {
		return ImmutableSet.copyOf(this);
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/collection/FreezableSet.kt
package dev.psygamer.wireframe.util.collection

import com.google.common.collect.ImmutableSet
import dev.psygamer.wireframe.util.IFreezable

abstract class FreezableSet<E> protected constructor(private val set: MutableSet<E>) : MutableSet<E>, IFreezable {

	@Volatile
	final override var frozen = false
		private set

	override val size
		get() = set.size

	override fun add(element: E): Boolean {
		IFreezable.throwIfFrozen(this)
		return set.add(element)
	}

	override fun addAll(elements: Collection<E>): Boolean {
		IFreezable.throwIfFrozen(this)
		return set.addAll(elements)
	}

	override fun remove(element: E): Boolean {
		IFreezable.throwIfFrozen(this)
		return set.remove(element)
	}

	override fun removeAll(elements: Collection<E>): Boolean {
		IFreezable.throwIfFrozen(this)
		return set.removeAll(elements.toSet())
	}

	override fun retainAll(elements: Collection<E>): Boolean {
		IFreezable.throwIfFrozen(this)
		return set.retainAll(elements.toSet())
	}

	override fun clear() {
		IFreezable.throwIfFrozen(this)
		set.clear()
	}

	override fun isEmpty() =
		set.isEmpty()

	override operator fun contains(element: E) =
		set.contains(element)

	override fun containsAll(elements: Collection<E>) =
		set.containsAll(elements)

	override operator fun iterator(): MutableIterator<E> =
		set.iterator()

	override fun freeze() {
		frozen = true
	}

	fun toImmutable(): ImmutableSet<E> {
		return ImmutableSet.copyOf(this)
	}
}